### PR TITLE
[MRG] expose uint16_codec in eeglab reader (+ minor fixes)

### PR DESCRIPTION
**Pull request by mmagnuski**
@agramfort 
Sorry it took so long, forgot about that. 

Now `read_epochs_eeglab` and `read_raw_eeglab` expose `uint16_codec`.
Additional changes:
* the warning when multiple events are found in an epoch - is now raised only once, not for each epoch.
* minor fixes to the docstrings - `verbose` arg didn't use `|` when listing possible input types
* fixes reading one channel set files

TODO:
- [x] test for `uint16_codec`
- [x] maybe a better description of that argument in the docstring - when it's useful etc.
- [x] update what's new
- [x] fix for reading one-channel files
- [x] test for reading one-channels files

### CODE REVIEW

File: mne/io/eeglab/eeglab.py [hunk 1]
```diff
@@ -467,6 +477,12 @@ def __init__(self, input_fname, events=None, event_id=None, tmin=0,
                 # invent event dict but use id > 0 so you know its a trigger
                 event_id = dict((ev, idx + 1) for idx, ev
                                 in enumerate(unique_ev))
+
+            # warn about multiple events in epoch if necessary
+            if warn_multiple_events:
+                warn('At least one epoch has multiple events. Only the latency'
+                     ' of the first event will be retained.')
```
Comment (jasmainak) [inline mne/io/eeglab/eeglab.py]: nice, thanks :)

File: mne/io/eeglab/eeglab.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def _get_info(eeg, montage, eog=()):
 
 def read_raw_eeglab(input_fname, montage=None, eog=(), event_id=None,
                     event_id_func='strip_to_integer', preload=False,
-                    verbose=None):
+                    verbose=None, uint16_codec=None):
```
Comment (jasmainak) [inline mne/io/eeglab/eeglab.py]: any chance of adding a test for this?

File: mne/io/eeglab/eeglab.py [hunk 3]
```diff
@@ -266,8 +271,10 @@ class RawEEGLAB(_BaseRaw):
         amount of memory). If preload is a string, preload is the file name of
         a memory-mapped file which is used to store the data on the hard
         drive (slower, requires less memory).
-    verbose : bool, str, int, or None
+    verbose : bool | str | int | None
         If not None, override default verbose level (see mne.verbose).
+    uint16_codec : str | None
+        mne.io.read_raw_eeglab(os.path.join(data_dir, set_fls[1]))
```
Comment (jasmainak) [inline mne/io/eeglab/eeglab.py]: docstring is cryptic

File: mne/io/eeglab/eeglab.py [hunk 4]
```diff
@@ -204,8 +206,10 @@ def read_epochs_eeglab(input_fname, events=None, event_id=None, montage=None,
         Names or indices of channels that should be designated EOG channels.
         If 'auto', the channel names containing ``EOG`` or ``EYE`` are used.
         Defaults to empty tuple.
-    verbose : bool, str, int, or None
+    verbose : bool | str | int | None
         If not None, override default verbose level (see mne.verbose).
+    uint16_codec : str | None
+        Codec to use for uint16 char arrays (defaults to system default codec).
```
Comment (jasmainak) [inline mne/io/eeglab/eeglab.py]: maybe add a comment as to how to know the system default codec?

File: mne/io/eeglab/eeglab.py [hunk 5]
```diff
@@ -285,14 +292,14 @@ class RawEEGLAB(_BaseRaw):
...
-                 verbose=None):
+                 verbose=None, uint16_codec=None):
         """Read EEGLAB .set file.
         """
         from scipy import io
         basedir = op.dirname(input_fname)
         _check_mat_struct(input_fname)
         eeg = io.loadmat(input_fname, struct_as_record=False,
-                         squeeze_me=True)['EEG']
+                         squeeze_me=True, uint16_codec=uint16_codec)['EEG']
```
Comment (jasmainak) [inline mne/io/eeglab/eeglab.py]: what version of `scipy` do you use? I don't have this argument? we might have to do some version checking for this ...

File: mne/io/eeglab/eeglab.py (hunk 2 above)
Comment (mmagnuski) [inline mne/io/eeglab/eeglab.py]: yes, it's in the todos :)

File: mne/io/eeglab/eeglab.py (hunk 3 above)
Comment (mmagnuski) [inline mne/io/eeglab/eeglab.py]: oops, something went wrong with my copy-paste kung-fu... :)

File: mne/io/eeglab/eeglab.py (hunk 4 above)
Comment (mmagnuski) [inline mne/io/eeglab/eeglab.py]: I think this docstring should change actually - I just copied it from scipy. `uint16_codec` is only useful when the set file you are trying to load has some non-ascii characters and default scipy.io.loadmat fails with buffer error. see https://github.com/scipy/scipy/issues/5516

File: mne/io/eeglab/eeglab.py (hunk 5 above)
Comment (mmagnuski) [inline mne/io/eeglab/eeglab.py]: I'm using 0.17.1 but that argument seems to be undocumented in that version (but it works)

Comment (mmagnuski) [inline mne/io/eeglab/eeglab.py]: (it goes to loadmat kwargs)

Comment (jasmainak) [inline mne/io/eeglab/eeglab.py]: yeah sounds good. I checked and I think this has been part of their machinery for a long time. So we're good here

**Comment by agramfort** [comment]
@jasmainak merge when you're happy

@mmagnuski don't forget the MRG status

**Comment by jasmainak** [comment]
@mmagnuski update `whats_new.rst` and you're ready for merge :)

File: mne/io/eeglab/tests/test_eeglab.py [hunk 6]
```diff
@@ -126,6 +126,6 @@ def test_io_set():
         warnings.simplefilter('always')
         assert_raises(NotImplementedError, read_epochs_eeglab,
                       bad_epochs_fname)
-    assert_equal(len(w), 3)
+    assert_equal(len(w), 1)
```
Comment (jasmainak) [inline mne/io/eeglab/tests/test_eeglab.py]: oh wait, I forgot you haven't updated the test yet ...

Comment (mmagnuski) [inline mne/io/eeglab/tests/test_eeglab.py]: Yeah, I'll add a test soon :)

**Comment by mmagnuski** [comment]
I'll change to MRG once all the checkboxes are marked :) This should happen before monday.

**Comment by mmagnuski** [comment]
I'm having problems programmatically creating a file that would test `uint16_codec` param. Maybe I'm doing something wrong, but I think if I successfully write a file with `io.savemat` - I will also be able to read it without problems.
But I have some problematic files on my computer (I'm only able to load these specifying `uint16_codec` - I could edit them in matlab to be really small (one-two channels, two samples etc.) and make sure they still cause the problem and upload them somewhere.
I'm also open to other ideas on how to test this.

**Comment by jasmainak** [comment]
you could make a PR here: https://github.com/mne-tools/mne-testing-data/tree/master/EEGLAB and add your problematic file here ... trying to make it as small is as possible by editing it on Matlab

**Comment by mmagnuski** [comment]
It turns out that if I edit and save the file on my computer the problem disappears - so I currently have no means to create a manageable set file that would reproduce the problem. :(
I think it is also a good indicator that maybe it is not worth having a strict tests - maybe I'll just add a line to tests that demonstrates that using `uint16_codec` does not break anything?

However - I was able to get the the files (set + fdt) size down to 471B (this is their summed size, but the fdt is just 8B), so I may look into reducting the size of current eeglab testing datasets (if you think it is worth doing).
Just for the record, removing all fields not used by mne eeglab reader helps:
```matlab
EEG = rmfield(EEG, {'setname', 'filename', 'filepath', 'subject', 'session', 'group', 'condition', 'comments', 'icaact', 'icawinv', 'icasphere', 'icaweights', 'icachansind', 'urchanlocs', 'chaninfo', 'ref', 'urevent', 'eventdescription', 'epochdescription', 'reject', 'stats', 'specdata', 'specicaact', 'splinefile', 'icasplinefile', 'dipfit', 'history', 'saved', 'etc', 'datfile'});
EEG.chanlocs = rmfield(EEG.chanlocs, {'theta', 'radius', 'sph_theta', 'sph_phi', 'sph_radius', 'type', 'urchan'});
```

**Comment by Eric89GXL** [comment]
It's probably okay to leave the test data as it is.

This is a corner case that we pretty much offload to `scipy`'s routines now, so I think it's okay not to have a great test (no need for new test data if it's not easy), although a smoke test that you can at least pass the parameter would indeed be good.

**Comment by mmagnuski** [comment]
Ah, and I discovered that trying to read one-channel set file, gives the same error as previously when reading a file with just one event. Is it worth fixing (if you think so I can submit a PR)?

**Comment by Eric89GXL** [comment]
Yeah, it can be part of this one if you want

**Comment by mmagnuski** [comment]
Ok, so I'll rebase and fix the one-chan issue.

**Comment by mmagnuski** [comment]
@jasmainak @Eric89GXL
All green, one last thing I could add is a test for one-channel epoched data, but I don't have to if you think it is good already.

**Comment by jasmainak** [comment]
looks clean @mmagnuski . Merging, thanks for the fix!

**Comment by mmagnuski** [comment]
Thanks!
